### PR TITLE
chore(macaroon): bump macaroon-bakery patch + refactor

### PR DESCRIPTION
**Pull request by jack-w-shaw**
The latest patch version of go-macaroon-bakery/macaroon-bakery contains a new constructor for RootKeyStores, which takes in a ContextBacking as a param.

This means we no longer need to implement stub methods for Backing.

Make use of this new constructor, remove all stub methods + casts, and make any refactors resulting from this

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- [x] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing
- [ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages

## QA steps

```
./main.sh -v cmr
```

Cherry pick the following commits to hotfix a known bug with external users and build juju:
- https://github.com/juju/juju/pull/17752/commits/12437cf19617af9808d546e8fc2a950bd95636e3

Copy your ec2 credential over to `./cred.yaml` such that:
```
cat ./cred.yaml
credentials:
  aws:
    ec2-robot-external:
      auth-type: access-key
      access-key: ...
      secret-key: ...
```

The run + observe:
```
$ juju bootstrap aws/eu-west-2 aws --config identity-url=https://api.jujucharms.com/identity --config allow-model-access=true
$ juju grant <lp user>@external superuser
$ juju change-user-password
$ juju logout
$ juju login --user <lp user>@external
Welcome, <lp user>@external. You are now logged into "lxd".

Current model set to "admin/controller".

$ juju whoami
Controller:  aws
Model:       admin/controller
User:        <lp user>@external

$ juju add-credential aws -f ./cred.yaml --controller aws --client
Using cloud "aws" from the controller to verify credentials.
Credential "ec2-robot-external" added locally for cloud "aws".

Controller credential "ec2-robot-external" for user "jack-shaw@external" for cloud "aws" on controller "aws" added.
For more information, see ‘juju show-credential aws ec2-robot-external’.

$ juju add-model m --credential ec2-robot-external
Uploading credential 'aws/jack-shaw@external/ec2-robot-external' to controller
Added 'm' model on aws/eu-west-2 with credential 'ec2-robot-external' for user 'jack-shaw'

$ juju deploy ubuntu
$ juju status
Model  Controller  Cloud/Region   Version      Timestamp
m      aws         aws/eu-west-2  4.0-beta4.1  14:46:11+01:00

App     Version  Status   Scale  Charm   Channel        Rev  Exposed  Message
ubuntu           waiting    0/1  ubuntu  latest/stable   24  no       waiting for machine

Unit      Workload  Agent       Machine  Public address  Ports  Message
ubuntu/0  waiting   allocating  0        18.134.209.114         waiting for machine

Machine  State    Address         Inst id              Base          AZ          Message
0        pending  18.134.209.114  i-0790d125c69f15f77  ubuntu@22.04  eu-west-2b  running
```

NOTE: There are still some bugs with external users. They will hopefully be resolved with https://github.com/juju/juju/pull/17752 + other work

### CODE REVIEW

File: apiserver/stateauthenticator/context_integration_test.go [hunk 1]
```diff
@@ -33,12 +33,12 @@ import (
 // code with the 'third-party' (Canonical controlled) go-macaroon-bakery code.
 //
 // Particularly much of go-macaroon-bakery depends on injecting an implementation
-// of an interface defined within the library (dbrootkeystore.Backing). The ways
-// in which this interface is then used is an implementation detal of the
+// of an interface defined within the library (dbrootkeystore.ContextBacking). The
+// ways in which this interface is then used is an implementation detail of the
```
Comment (Aflynn50) [inline apiserver/stateauthenticator/context_integration_test.go:37]: The second sentence here seems grammatically incorrect. "ways" and "is" should be "way" and "is" or "ways" and "are".

**Review by Aflynn50 — APPROVED** [review]
Looks good.

File: apiserver/stateauthenticator/context_integration_test.go (hunk 1 above)
Comment (jack-w-shaw) [reply apiserver/stateauthenticator/context_integration_test.go]: The plural "ways" refers to the multiple ways the interface can be used, the singular "is" refers to the singular interface. So I _think_ it's technically ok

However, it is a little unclear to read, so I've amended

**Pull request merged**